src/lock_free/mwcas_descriptor.cpp

```diff
@@ -0,0 +1,191 @@
+ * Local constants
+ *############################################################################*/
+
+/// @brief An offset for right-shifting to extract the "begin position".
+constexpr uint64_t kPosShift = 47;
+
+/// @brief An offset for right-shifting to extract the "reference counter".
+constexpr uint64_t kCntShift = 50;
+
+/// @brief A constant for incrementing the "reference counter".
+constexpr uint64_t kCntUnit = 1UL << kCntShift;
+
+/// @brief A bitmask with only the "begin position" portion set to 1.
+constexpr uint64_t kPosMask = (((1UL << (1 + kCntShift - kPosShift)) - 1) << kPosShift);
+
+/// @brief A bitmask with only the "reference counter" portion set to 1.
+constexpr uint64_t kCntMask = (((1UL << (64 - kCntShift)) - 1) << kCntShift);
+
+/// @brief A bitmask with only the "descriptor address" portion set to 1.
+constexpr uint64_t kAddrMask = (1UL << 47) - 1;
+
+/// @brief A bitmask with only the "MwCAS FLAG" and "descriptor address" portions set to 1.
+constexpr uint64_t kDescMask = kMwCASFlag | kAddrMask;
+}  // namespace
+
+/*##############################################################################
+ * Static utilities
+ *############################################################################*/
+
+auto
+MwCASDescriptor::GetDescriptor()  //
+    -> MwCASDescriptor *
+{
+  auto *page = tls.release();
+  auto *desc = (page == nullptr) ? new MwCASDescriptor{} : static_cast<MwCASDescriptor *>(page);
+  desc->target_cnt_ = 0;
+  return desc;
+}
+
+/*##############################################################################
+ * Utilities
+ *############################################################################*/
+auto
+MwCASDescriptor::MwCAS()  //
+    -> bool
+{
+  // set a memory fence
+  stat_.store(kUndecided, kRelease);
+  const auto succeeded = MwCASInternal();
+  return succeeded;
+}
+
+auto
+MwCASDescriptor::MwCASInternal(  //
+    const size_t begin_pos       //
+    ) -> bool
+{
+  const auto desc_addr = std::bit_cast<uint64_t>(this) | kMwCASFlag;
+  auto cur_stat = stat_.load(kSeqCst);
+  if (cur_stat == kUndecided) {
+    auto stat = kSucceeded;
+    for (size_t i = begin_pos; i < target_cnt_; ++i) {
+      auto &target = targets_[i];
+      if (!EmbedDescriptor(desc_addr, i)) {
+        stat = kFailed;
+        break;
+      }
+    }
+    cur_stat = stat_.load(kSeqCst);
+    if (cur_stat == kUndecided && stat_.compare_exchange_strong(cur_stat, stat, kSeqCst, kSeqCst)) {
+      cur_stat = stat;
+    }
+  }
+
+  auto ret = (cur_stat == kSucceeded);
+  auto follow_cnt = 0;
+  if (ret) {
+    for (size_t i = 0; i < target_cnt_; ++i) {
+      auto &target = targets_[i];
+      auto expected = target.addr->load(kSeqCst);
+      if (((expected ^ desc_addr) & kDescMask) == 0UL) {
+        target.addr->compare_exchange_strong(expected, target.new_val, target.fence, target.fence);
+        follow_cnt += ((expected & kCntMask) >> kCntShift);
+      } else {
+        follow_cnt += target.cnt;
+      }
```

Comment:
ここの手順として， **埋め込まれた記述子を交換する前に記述子側のカウンタを加算する** 必要があります．そうしないと複数のスレッドが追従処理を行った場合，つまり3つ以上のスレッドでこのfinalize処理を行う場合に，最後のスレッド（記述子を交換したスレッドでもこの対象領域以外から追従処理を開始したスレッド）が正しいカウンタを観測できない可能性があります．

ちゃんと検証してませんが，イメージとしては以下のような感じです．

```suggestion
      while (true) {
        auto expected = target.addr->load(kSeqCst);
        auto cur_cnt = target.cnt.load(kSeqCst);
        auto cnt = static_cast<uint32_t>((expected & kCntMask) >> kCntShift);
        if (((expected ^ desc_addr) & kDescMask)  // already swapped
            || (cur_cnt == cnt && target.addr->compare_exchange_strong(
                  expected, target.new_val, target.fence, target.fence))) {  // swap succeeded
          follow_cnt += cur_cnt;
          break;
        }

        // the descriptor's counter should be updated
        while (cur_cnt < cnt) {
          if (target.cnt.compare_exchange_weak(cur_cnt, cnt, kSeqCst, kSeqCst)) break;
          CPP_UTILITY_SPINLOCK_HINT
        }
      }
```

が，見ての通りだいぶ長いので，これも関数化してしまったほうがよい気がします．引数で `target.old_val` か `target.new_val` をもらうようにしておけば，redo/undoどちらのパターンでも使い回せるはずです．